File: app/src/main/java/com/example/appv1/paciente/HomePaciente.java
package com.example.appv1.paciente;
import android.Manifest;
import android.annotation.SuppressLint;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import com.example.appv1.R;

import java.io.IOException;
import java.io.InputStream;
import java.util.UUID;

public class HomePaciente extends AppCompatActivity {
    private BluetoothAdapter bluetoothAdapter;
    private BluetoothSocket btSocket;
    private InputStream inputStream;
    private Handler handler;

    private TextView temperaturaView, pulsoView, oxigenoView, estadoConexionView;

    private final String DEVICE_NAME = "ESP32_HealthMonitor"; // Nombre del Bluetooth del ESP32
    private final UUID BT_UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");

    private static final int REQUEST_BLUETOOTH_PERMISSION = 1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home_paciente); // Usa la vista que pasaste

        temperaturaView = findViewById(R.id.temperatura);
        pulsoView = findViewById(R.id.hrtpulso);
        oxigenoView = findViewById(R.id.oxigeno);
        estadoConexionView = findViewById(R.id.txtEstadoConexion);

        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        handler = new Handler(Looper.getMainLooper());

        // Pedir permisos si es necesario
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            requestBluetoothPermissions();
        } else {
            connectToESP32();
        }
    }

    private void requestBluetoothPermissions() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.BLUETOOTH_CONNECT) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.BLUETOOTH_CONNECT}, REQUEST_BLUETOOTH_PERMISSION);
        } else {
            connectToESP32(); // Si ya tiene permisos, conectar al ESP32
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_BLUETOOTH_PERMISSION) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                connectToESP32();
            } else {
                Toast.makeText(this, "Permiso de Bluetooth denegado", Toast.LENGTH_SHORT).show();
            }
        }
    }

    @SuppressLint("MissingPermission")
    private void connectToESP32() {
        if (bluetoothAdapter == null || !bluetoothAdapter.isEnabled()) {
            Toast.makeText(this, "Bluetooth no disponible o apagado", Toast.LENGTH_SHORT).show();
            estadoConexionView.setText("Desconectado");
            estadoConexionView.setTextColor(getResources().getColor(android.R.color.holo_red_light));
            return;
        }

        for (BluetoothDevice device : bluetoothAdapter.getBondedDevices()) {
            if (device.getName().equals(DEVICE_NAME)) {
                try {
                    btSocket = device.createRfcommSocketToServiceRecord(BT_UUID);
                    btSocket.connect();
                    inputStream = btSocket.getInputStream();
                    estadoConexionView.setText("Conectado");
                    estadoConexionView.setTextColor(getResources().getColor(android.R.color.holo_green_light));
                    readData();
                    return;
                } catch (IOException e) {
                    Toast.makeText(this, "Error al conectar", Toast.LENGTH_SHORT).show();
                    estadoConexionView.setText("Desconectado");
                    estadoConexionView.setTextColor(getResources().getColor(android.R.color.holo_red_light));
                }
            }
        }
    }

    private void readData() {
        new Thread(() -> {
            byte[] buffer = new byte[1024];
            int bytes;

            while (true) {
                try {
                    if (inputStream.available() > 0) {
                        bytes = inputStream.read(buffer);
                        String data = new String(buffer, 0, bytes);
                        updateUI(data);
                    }
                } catch (IOException e) {
                    handler.post(() -> {
                        estadoConexionView.setText("Desconectado");
                        estadoConexionView.setTextColor(getResources().getColor(android.R.color.holo_red_light));
                    });
                    break;
                }
            }
        }).start();
    }

    private void updateUI(String data) {
        handler.post(() -> {
            try {
                String[] values = data.split(" ");

                    temperaturaView.setText(values[0]); // Ej: Temp:36.5
                    pulsoView.setText(values[1]); // Ej: HR:78
                    oxigenoView.setText(values[2]); // Ej: SpO2:98


            } catch (Exception e) {
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        try {
            if (btSocket != null) {
                btSocket.close();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/example/appv1/paciente/HomePaciente.kt
package com.example.appv1.paciente

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresPermission
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatButton
import androidx.core.content.ContextCompat
import com.example.appv1.R
import com.example.appv1.paciente.service.BlePacienteService

class HomePaciente : AppCompatActivity() {

    private lateinit var pulsoView: TextView
    private lateinit var oxigenoView: TextView
    private lateinit var temperaturaView: TextView
    private lateinit var estadoConexionView: TextView
    private lateinit var btnMedir: AppCompatButton

    private val receptor = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            val pulso = intent?.getIntExtra("pulso", 0) ?: 0
            val oxigeno = intent?.getIntExtra("oxigeno", 0) ?: 0
            val temperatura = intent?.getDoubleExtra("temperatura", 0.0) ?: 0.0

            pulsoView.text = pulso.toString()
            oxigenoView.text = oxigeno.toString()
            temperaturaView.text = "$temperatura °C"
        }
    }

    @SuppressLint("UnspecifiedRegisterReceiverFlag")
    override fun onResume() {
        super.onResume()

        // Mostrar como desconectado por defecto
        estadoConexionView.text = "Desconectado"
        estadoConexionView.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_light))

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(receptor, IntentFilter("DATOS_PULSERA"), Context.RECEIVER_NOT_EXPORTED)
        } else {
            registerReceiver(receptor, IntentFilter("DATOS_PULSERA"))
        }
    }

    override fun onPause() {
        super.onPause()
        unregisterReceiver(receptor)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home_paciente)

        pulsoView = findViewById(R.id.hrtpulso)
        oxigenoView = findViewById(R.id.oxigeno)
        temperaturaView = findViewById(R.id.temperatura)
        estadoConexionView = findViewById(R.id.txtEstadoConexion)
        btnMedir = findViewById(R.id.button2)

        // Valor por defecto al iniciar
        estadoConexionView.text = "Desconectado"
        estadoConexionView.setTextColor(ContextCompat.getColor(this, android.R.color.holo_red_light))

        btnMedir.setOnClickListener {
            pedirPermisos()
        }
    }

    private fun pedirPermisos() {
        val permisos = mutableListOf(
            Manifest.permission.BLUETOOTH,
            Manifest.permission.BLUETOOTH_ADMIN,
            Manifest.permission.ACCESS_FINE_LOCATION
        )

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            permisos.add(Manifest.permission.BLUETOOTH_SCAN)
            permisos.add(Manifest.permission.BLUETOOTH_CONNECT)
        }

        permisosLauncher.launch(permisos.toTypedArray())
    }

    @SuppressLint("MissingPermission")
    private val permisosLauncher = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { resultado ->
        val todosOtorgados = resultado.entries.all { it.value }

        if (!todosOtorgados) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                if (ContextCompat.checkSelfPermission(
                        this,
                        Manifest.permission.BLUETOOTH_SCAN
                    ) == PackageManager.PERMISSION_GRANTED
                ) {
                    escanearYConectar()
                } else {
                    Toast.makeText(this, "Permiso BLUETOOTH_SCAN no otorgado", Toast.LENGTH_SHORT).show()
                }
            } else {
                escanearYConectar()
            }
        } else {
            Toast.makeText(this, "Debes aceptar todos los permisos para conectar la pulsera", Toast.LENGTH_LONG).show()
        }
    }

    @RequiresPermission(Manifest.permission.BLUETOOTH_SCAN)
    private fun escanearYConectar() {
        val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        val scanner = bluetoothAdapter.bluetoothLeScanner

        val filtro = ScanFilter.Builder()
            .setDeviceName("SCI-BAND paciente")
            .build()

        val settings = ScanSettings.Builder().build()

        scanner.startScan(listOf(filtro), settings, object : ScanCallback() {
            @RequiresPermission(Manifest.permission.BLUETOOTH_SCAN)
            override fun onScanResult(callbackType: Int, result: ScanResult) {
                scanner.stopScan(this)

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                    if (!this@HomePaciente.isInForeground()) {
                        Toast.makeText(this@HomePaciente, "No se puede iniciar el servicio si la app no está activa", Toast.LENGTH_SHORT).show()
                        return
                    }
                }

                val intent = Intent(this@HomePaciente, BlePacienteService::class.java)
                intent.putExtra("device", result.device)

                ContextCompat.startForegroundService(this@HomePaciente, intent)
                Toast.makeText(this@HomePaciente, "Pulsera conectada", Toast.LENGTH_SHORT).show()


                estadoConexionView.text = "Conectado"
                estadoConexionView.setTextColor(ContextCompat.getColor(this@HomePaciente, android.R.color.holo_green_light))
            }

            override fun onScanFailed(errorCode: Int) {
                Toast.makeText(this@HomePaciente, "Fallo al escanear BLE ($errorCode)", Toast.LENGTH_SHORT).show()
                Log.e("BLE", "Scan falló: $errorCode")


                estadoConexionView.text = "Desconectado"
                estadoConexionView.setTextColor(ContextCompat.getColor(this@HomePaciente, android.R.color.holo_red_light))
            }

            fun Context.isInForeground(): Boolean {
                val activityManager = getSystemService(Context.ACTIVITY_SERVICE) as android.app.ActivityManager
                val appProcesses = activityManager.runningAppProcesses ?: return false
                val packageName = packageName
                for (appProcess in appProcesses) {
                    if (appProcess.importance == android.app.ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND &&
                        appProcess.processName == packageName) {
                        return true
                    }
                }
                return false
            }
        })
    }
}
